File: import-to-contentful.py
#!/usr/bin/env python3

"""
Import CSV data into Contentful using Management API

This script:
1. Reads contentful-import.csv
2. Uploads images to Contentful Media library
3. Creates Tip entries with all fields
4. Converts plain text body to Rich Text format
5. Publishes all entries

Requirements:
- pip install contentful-management
- Management API token (different from Delivery API token)

Usage:
  CONTENTFUL_MANAGEMENT_TOKEN=your-token python3 import-to-contentful.py
"""

import os
import sys
import csv
import time
import base64
from pathlib import Path

try:
    from contentful_management import Client
except ImportError:
    print('❌ Error: contentful-management package not found.')
    print('Install it with: pip install contentful-management')
    sys.exit(1)

space_id = os.getenv('CONTENTFUL_SPACE_ID', 'r4mpogvp9tf2')
management_token = os.getenv('CONTENTFUL_MANAGEMENT_TOKEN')

if not management_token:
    print('❌ Error: CONTENTFUL_MANAGEMENT_TOKEN environment variable is required')
    print('\nTo get a Management API token:')
    print('1. Go to Contentful → Settings → API keys')
    print('2. Click "Content Management API - access token"')
    print('3. Create a new token or use an existing one')
    print('\nThen run:')
    print('  CONTENTFUL_MANAGEMENT_TOKEN=your-token python3 import-to-contentful.py')
    sys.exit(1)

CSV_FILE = Path(__file__).parent / 'contentful-import.csv'
client = Client(management_token)

def text_to_rich_text(text):
    """Convert plain text to Contentful Rich Text Document"""
    if not text or not text.strip():
        return {
            'nodeType': 'document',
            'data': {},
            'content': []
        }
    
    # Split into paragraphs (double newlines)
    paragraphs = [p.strip() for p in text.split('\n\n') if p.strip()]
    
    if not paragraphs:
        # Single paragraph - split by single newlines
        lines = [line.strip() for line in text.split('\n') if line.strip()]
        return {
            'nodeType': 'document',
            'data': {},
            'content': [{
                'nodeType': 'paragraph',
                'data': {},
                'content': [
                    {
                        'nodeType': 'text',
                        'value': line or ' ',
                        'marks': [],
                        'data': {}
                    }
                    for line in lines
                ]
            }]
        }
    
    # Multiple paragraphs
    return {
        'nodeType': 'document',
        'data': {},
        'content': [{
            'nodeType': 'paragraph',
            'data': {},
            'content': [
                {
                    'nodeType': 'text',
                    'value': line or ' ',
                    'marks': [],
                    'data': {}
                }
                for line in para.split('\n')
            ]
        } for para in paragraphs]
    }

def get_content_type(filename):
    """Get MIME type from filename"""
    ext = Path(filename).suffix.lower()
    types = {
        '.jpg': 'image/jpeg',
        '.jpeg': 'image/jpeg',
        '.png': 'image/png',
        '.gif': 'image/gif',
        '.webp': 'image/webp'
    }
    return types.get(ext, 'image/jpeg')

async def upload_image(space, environment, image_path):
    """Upload image file to Contentful"""
    if not image_path or not image_path.strip():
        return None
    
    full_path = Path(__file__).parent / image_path
    if not full_path.exists():
        print(f'  ⚠️  Image not found: {image_path}')
        return None
    
    try:
        filename = full_path.name
        content_type = get_content_type(filename)
        
        # Read file
        with open(full_path, 'rb') as f:
            file_data = f.read()
        
        # Create asset
        asset = environment.assets().create({
            'fields': {
                'title': {
                    'en-US': filename
                },
                'file': {
                    'en-US': {
                        'contentType': content_type,
                        'fileName': filename,
                        'upload': base64.b64encode(file_data).decode('utf-8')
                    }
                }
            }
        })
        
        # Process asset
        asset.process()
        
        # Wait for processing
        processed = False
        attempts = 0
        while not processed and attempts < 10:
            time.sleep(1)
            updated_asset = environment.assets().find(asset.id)
            if updated_asset.fields.get('file'):
                processed = True
            attempts += 1
        
        # Publish asset
        asset.publish()
        
        print(f'  ✓ Uploaded image: {filename}')
        return asset.id
    except Exception as error:
        print(f'  ✗ Failed to upload {image_path}: {error}')
        return None

def read_csv():
    """Read CSV file"""
    rows = []
    with open(CSV_FILE, 'r', encoding='utf-8') as f:
        reader = csv.DictReader(f)
        for row in reader:
            rows.append(row)
    return rows

async def import_to_contentful():
    """Main import function"""
    try:
        print('📖 Reading CSV file...')
        rows = read_csv()
        print(f'✅ Found {len(rows)} tips to import\n')
        
        print('🔌 Connecting to Contentful...')
        space = client.spaces().find(space_id)
        environment = space.environments().find('master')
        print('✅ Connected\n')
        
        results = {
            'success': 0,
            'failed': 0,
            'skipped': 0
        }
        
        for i, row in enumerate(rows, 1):
            tip_number = int(row['tipNumber'])
            print(f'[{i}/{len(rows)}] Importing tip {tip_number}: {row["title"]}')
            
            try:
                # Check if entry already exists
                existing = environment.entries().all({
                    'content_type': 'tip',
                    'fields.slug': row['slug'],
                    'limit': 1
                })
                
                if existing:
                    print(f'  ⚠️  Entry with slug "{row["slug"]}" already exists, skipping...')
                    results['skipped'] += 1
                    continue
                
                # Upload image if provided
                image_asset_id = None
                if row.get('imagePath') and row['imagePath'].strip():
                    image_asset_id = await upload_image(space, environment, row['imagePath'])
                
                # Convert body to Rich Text
                rich_text_body = text_to_rich_text(row['body'])
                
                # Create entry
                entry_data = {
                    'fields': {
                        'title': {'en-US': row['title']},
                        'slug': {'en-US': row['slug']},
                        'tipNumber': {'en-US': tip_number},
                        'body': {'en-US': rich_text_body}
                    }
                }
                
                if image_asset_id:
                    entry_data['fields']['image'] = {
                        'en-US': {
                            'sys': {
                                'type': 'Link',
                                'linkType': 'Asset',
                                'id': image_asset_id
                            }
                        }
                    }
                
                entry = environment.entries().create('tip', entry_data)
                
                # Publish entry
                entry.publish()
                
                print(f'  ✅ Created and published: {row["slug"]}')
                results['success'] += 1
                
                # Small delay to avoid rate limits
                time.sleep(0.5)
            except Exception as error:
                print(f'  ✗ Failed: {error}')
                results['failed'] += 1
        
        print('\n📊 Import Summary:')
        print(f'   ✅ Success: {results["success"]}')
        print(f'   ⚠️  Skipped: {results["skipped"]}')
        print(f'   ✗ Failed: {results["failed"]}')
        print('\n🎉 Import complete!')
        
    except Exception as error:
        print(f'❌ Import failed: {error}')
        if '401' in str(error):
            print('\nThe Management API token is invalid or expired.')
        elif '404' in str(error):
            print('\nContent type "tip" not found. Make sure it exists in your space.')
        sys.exit(1)

if __name__ == '__main__':
    import asyncio
    asyncio.run(import_to_contentful())

File: import-to-contentful.js
#!/usr/bin/env node

/**
 * Import CSV data into Contentful using Management API
 * 
 * This script:
 * 1. Reads contentful-import.csv
 * 2. Uploads images to Contentful Media library
 * 3. Creates Tip entries with all fields
 * 4. Converts plain text body to Rich Text format
 * 5. Publishes all entries
 * 
 * Requirements:
 * - npm install contentful-management
 * - Management API token (different from Delivery API token)
 * 
 * Usage:
 *   CONTENTFUL_MANAGEMENT_TOKEN=your-token node import-to-contentful.js
 */

const contentfulManagement = require('contentful-management');
const fs = require('fs');
const path = require('path');
const csv = require('csv-parser');

const spaceId = process.env.CONTENTFUL_SPACE_ID || 'r4mpogvp9tf2';
const managementToken = process.env.CONTENTFUL_MANAGEMENT_TOKEN;

if (!managementToken) {
  console.error('❌ Error: CONTENTFUL_MANAGEMENT_TOKEN environment variable is required');
  console.error('\nTo get a Management API token:');
  console.error('1. Go to Contentful → Settings → API keys');
  console.error('2. Click "Content Management API - access token"');
  console.error('3. Create a new token or use an existing one');
  console.error('\nThen run:');
  console.error(`  CONTENTFUL_MANAGEMENT_TOKEN=your-token node import-to-contentful.js`);
  process.exit(1);
}

const CSV_FILE = path.join(__dirname, 'contentful-import.csv');
const client = contentfulManagement.createClient({
  accessToken: managementToken,
});

/**
 * Convert plain text to Contentful Rich Text Document
 */
function textToRichText(text) {
  if (!text || !text.trim()) {
    return {
      nodeType: 'document',
      data: {},
      content: []
    };
  }

  // Split text into paragraphs (double newlines)
  const paragraphs = text.split(/\n\n+/).filter(p => p.trim());
  
  if (paragraphs.length === 0) {
    // Single paragraph
    return {
      nodeType: 'document',
      data: {},
      content: [
        {
          nodeType: 'paragraph',
          data: {},
          content: text.split('\n').map(line => ({
            nodeType: 'text',
            value: line.trim() || ' ',
            marks: [],
            data: {}
          }))
        }
      ]
    };
  }

  // Multiple paragraphs
  return {
    nodeType: 'document',
    data: {},
    content: paragraphs.map(para => ({
      nodeType: 'paragraph',
      data: {},
      content: para.split('\n').map(line => ({
        nodeType: 'text',
        value: line.trim() || ' ',
        marks: [],
        data: {}
      }))
    }))
  };
}

/**
 * Upload image file to Contentful
 * 
 * Uses createAssetFromFiles which handles the upload process automatically
 */
async function uploadImage(space, environment, imagePath) {
  if (!imagePath || !imagePath.trim()) {
    return null;
  }

  const fullPath = path.join(__dirname, imagePath);
  if (!fs.existsSync(fullPath)) {
    console.warn(`⚠️  Image not found: ${imagePath}`);
    return null;
  }

  try {
    const fileBuffer = fs.readFileSync(fullPath);
    const fileName = path.basename(fullPath);
    const contentType = getContentType(fileName);

    // Create asset from file buffer
    const asset = await environment.createAssetFromFiles({
      fields: {
        title: {
          'en-US': fileName
        },
        file: {
          'en-US': {
            contentType: contentType,
            fileName: fileName,
            file: fileBuffer
          }
        }
      }
    });

    // Process asset
    await asset.processForAllLocales();
    
    // Wait for processing to complete
    let processed = false;
    let attempts = 0;
    while (!processed && attempts < 10) {
      await new Promise(resolve => setTimeout(resolve, 1000));
      const updatedAsset = await environment.getAsset(asset.sys.id);
      if (updatedAsset.fields.file && updatedAsset.fields.file['en-US']) {
        processed = true;
      }
      attempts++;
    }

    // Publish asset
    await asset.publish();

    console.log(`  ✓ Uploaded image: ${fileName}`);
    return asset.sys.id;
  } catch (error) {
    console.error(`  ✗ Failed to upload ${imagePath}:`, error.message);
    return null;
  }
}

/**
 * Get content type from filename
 */
function getContentType(filename) {
  const ext = path.extname(filename).toLowerCase();
  const types = {
    '.jpg': 'image/jpeg',
    '.jpeg': 'image/jpeg',
    '.png': 'image/png',
    '.gif': 'image/gif',
    '.webp': 'image/webp'
  };
  return types[ext] || 'image/jpeg';
}

/**
 * Read CSV file
 */
function readCSV() {
  return new Promise((resolve, reject) => {
    const rows = [];
    fs.createReadStream(CSV_FILE)
      .pipe(csv())
      .on('data', (row) => rows.push(row))
      .on('end', () => resolve(rows))
      .on('error', reject);
  });
}

/**
 * Main import function
 */
async function importToContentful() {
  try {
    console.log('📖 Reading CSV file...');
    const rows = await readCSV();
    console.log(`✅ Found ${rows.length} tips to import\n`);

    console.log('🔌 Connecting to Contentful...');
    const space = await client.getSpace(spaceId);
    const environment = await space.getEnvironment('master');
    console.log('✅ Connected\n');

    const results = {
      success: 0,
      failed: 0,
      skipped: 0
    };

    for (let i = 0; i < rows.length; i++) {
      const row = rows[i];
      const tipNumber = parseInt(row.tipNumber, 10);
      
      console.log(`[${i + 1}/${rows.length}] Importing tip ${tipNumber}: ${row.title}`);

      try {
        // Check if entry already exists
        const existing = await environment.getEntries({
          content_type: 'tip',
          'fields.slug': row.slug,
          limit: 1
        });

        if (existing.items.length > 0) {
          console.log(`  ⚠️  Entry with slug "${row.slug}" already exists, skipping...`);
          results.skipped++;
          continue;
        }

        // Upload image if provided
        let imageAssetId = null;
        if (row.imagePath && row.imagePath.trim()) {
          imageAssetId = await uploadImage(space, environment, row.imagePath);
        }

        // Convert body text to Rich Text
        const richTextBody = textToRichText(row.body);

        // Create entry
        const entry = await environment.createEntry('tip', {
          fields: {
            title: {
              'en-US': row.title
            },
            slug: {
              'en-US': row.slug
            },
            tipNumber: {
              'en-US': tipNumber
            },
            body: {
              'en-US': richTextBody
            },
            ...(imageAssetId && {
              image: {
                'en-US': {
                  sys: {
                    type: 'Link',
                    linkType: 'Asset',
                    id: imageAssetId
                  }
                }
              }
            })
          }
        });

        // Publish entry
        await entry.publish();

        console.log(`  ✅ Created and published: ${row.slug}`);
        results.success++;

        // Small delay to avoid rate limits
        await new Promise(resolve => setTimeout(resolve, 500));
      } catch (error) {
        console.error(`  ✗ Failed: ${error.message}`);
        results.failed++;
      }
    }

    console.log('\n📊 Import Summary:');
    console.log(`   ✅ Success: ${results.success}`);
    console.log(`   ⚠️  Skipped: ${results.skipped}`);
    console.log(`   ✗ Failed: ${results.failed}`);
    console.log(`\n🎉 Import complete!`);

  } catch (error) {
    console.error('❌ Import failed:', error.message);
    if (error.message.includes('401')) {
      console.error('\nThe Management API token is invalid or expired.');
    } else if (error.message.includes('404')) {
      console.error('\nContent type "tip" not found. Make sure it exists in your space.');
    }
    process.exit(1);
  }
}

// Check if contentful-management is installed
try {
  require.resolve('contentful-management');
  importToContentful();
} catch (e) {
  console.error('❌ Error: contentful-management package not found.');
  console.error('Install it with: npm install contentful-management');
  console.error('\nYou may need to install it in the worktree directory:');
  console.error('  cd .worktrees/001-contentful-migration');
  console.error('  npm install contentful-management');
  process.exit(1);
}

